mlapi/holiday.go

```diff
@@ -0,0 +1,129 @@
+	// of the holiday.
+	ICalURL *string `json:"iCalUrl,omitempty"`
+	// ICalTimeZone is the timezone to use for 'All Day' events on the iCal in ICalURL,
+	// if present. This is required because All Day events don't come with time zone information by default,
+	// and there is no well-adhered-to standard for timezones of entire iCal files.
+	ICalTimeZone *string `json:"iCalTimeZone,omitempty"`
+
+	// CustomPeriods are holiday periods that are specified explicitly.
+	CustomPeriods CustomPeriods `json:"customPeriods,omitempty"`
+
+	// Jobs is a slice of IDs of Jobs that are using this holiday.
+	// Requests may specify either IDs or names. Responses will always contain IDs.
+	Jobs []string `json:"jobs"`
+}
+
+// CustomPeriods is a slice of CustomPeriods representing all occurrences
+// of a holiday.
+type CustomPeriods []CustomPeriod
+
+// CustomPeriod is a single period to be included in a holiday.
+type CustomPeriod struct {
+	// Name is the name of this period.
+	Name string `json:"name"`
+	// StartTime is the (inclusive) start of this period.
+	StartTime time.Time `json:"startTime"`
+	// EndTime is the (exclusive) end of this period.
+	EndTime time.Time `json:"endTime"`
+}
+
+type holidayResponseWrapper struct {
+	Status string  `json:"status"`
+	Data   Holiday `json:"data"`
+	Error  string  `json:"error"`
+}
+
+// NewHoliday creates a new holiday.
+func (c *Client) NewHoliday(ctx context.Context, holiday Holiday) (Holiday, error) {
+	data, err := json.Marshal(holiday)
+	if err != nil {
+		return Holiday{}, err
+	}
+	result := holidayResponseWrapper{}
+	err = c.request(ctx, "POST", "/manage/api/v1/holidays", nil, bytes.NewBuffer(data), &result)
+	if err != nil {
+		return Holiday{}, err
+	}
+	return result.Data, err
+}
+
+// Holiday fetches an existing holiday.
+func (c *Client) Holiday(ctx context.Context, id string) (Holiday, error) {
+	result := holidayResponseWrapper{}
+	err := c.request(ctx, "GET", "/manage/api/v1/holidays/"+id, nil, nil, &result)
+	if err != nil {
+		return Holiday{}, err
+	}
+	return result.Data, err
+}
+
+// UpdateHoliday updates an existing holiday.
+func (c *Client) UpdateHoliday(ctx context.Context, holiday Holiday) (Holiday, error) {
+	id := holiday.ID
+	// Clear the ID before sending otherwise validation fails.
+	holiday.ID = ""
+	data, err := json.Marshal(holiday)
+	if err != nil {
+		return Holiday{}, err
+	}
+
+	result := holidayResponseWrapper{}
+	err = c.request(ctx, "POST", "/manage/api/v1/holidays/"+id, nil, bytes.NewBuffer(data), &result)
+	if err != nil {
+		return Holiday{}, err
+	}
+	return result.Data, err
+}
+
+// DeleteHoliday deletes an existing holiday.
+func (c *Client) DeleteHoliday(ctx context.Context, id string) error {
+	return c.request(ctx, "DELETE", "/manage/api/v1/holidays/"+id, nil, nil, nil)
+}
+
+// LinkJobHolidays links a job to a set of holidays.
+// Only the ID and Holidays fields of the Job struct are used.
+func (c *Client) LinkJobHolidays(ctx context.Context, job Job) (Job, error) {
```

Comment:
A few thoughts on this function:
1. It should probably be in `job.go` as it is calling `.../jobs/...`
2. `LinkJobHolidays` sounds a bit awkward to me, maybe just `LinkHolidays` or `LinkHolidaysToJob`?
3. Rather than take a `Job` as an argument and then do some weird clearing, maybe do `jobID uuid.UUID, holidays []string`? I think that might be a bit more ergonomic/straightforward for users.

Reply:
👍 those all sound sensible to me.